### PR TITLE
ddtrace/tracer: support for DD_TRACE_ENABLED environment variable

### LINKED ISSUE #946: proposal: Support for DD_TRACE_ENABLED environment variable
Many of the DataDog clients support the `DD_TRACE_ENABLED` environment variable. For example:
- Java: https://docs.datadoghq.com/tracing/setup_overview/setup/java/?tab=containers
- Node: https://docs.datadoghq.com/tracing/setup_overview/setup/nodejs/?tab=containers
- Python: https://ddtrace.readthedocs.io/en/stable/configuration.html
- Go Lambdas: https://github.com/DataDog/datadog-lambda-go

It would be nice if this library supported it in the same manner so that consumers could leverage the built-in configuration environment variables to toggle tracing on and off, as opposed to having to instrument custom logic within the application to control it.

### PR DESCRIPTION
**Pull request by AHHussain569**
* Allow for tracing to be enabled based on an environment variable or start option
* When disabled, tracer is replaced on Start with NoopTracer

resolves #946

### CODE REVIEW

**Comment by AHHussain569** [comment]
Could someone with permissions please re-run the pipeline, the specific failing tests are unrelated to the code changes made and pass locally

**Review by gbbr — DISMISSED** [review]
Please take a moment to read our contribution guidelines...

Lastly, why not do this in your own program?
```go
if ok, err := strconv.ParseBool(os.Getenv("DD_TRACE_ENABLED")); ok && err == nil {
    tracer.Start(...)
    defer tracer.Stop()
}
```

No need to add this into the library.

**Comment by knusbaum** [comment]
Hey, @gbbr. I approved this change in #946, since this is a feature present in many other tracers and should be a small code change.
Is this OK with you, or do we need to revisit the proposal?

**Comment by knusbaum** [comment]
I've merged @gbbr's suggestions and this can be merged once the tests pass.

File: ddtrace/tracer/option.go [hunk 1]
```diff
@@ -194,6 +197,8 @@ func newConfig(opts ...StartOption) *config {
 	c.logStartup = internal.BoolEnv("DD_TRACE_STARTUP_LOGS", true)
 	c.runtimeMetrics = internal.BoolEnv("DD_RUNTIME_METRICS_ENABLED", false)
 	c.debug = internal.BoolEnv("DD_TRACE_DEBUG", false)
+	c.traceEnabled = internal.BoolEnv("DD_TRACE_ENABLED", true)
```
Comment (knusbaum) [inline ddtrace/tracer/option.go:200]:
```suggestion
	c.enabled = internal.BoolEnv("DD_TRACE_ENABLED", true)
```

File: ddtrace/tracer/option.go [hunk 2]
```diff
@@ -481,6 +486,13 @@ func WithHostname(name string) StartOption {
 	}
 }
 
+// WithTraceEnabled allows specifying whether tracing will be enabled
+func WithTraceEnabled(traceEnabled bool) StartOption {
+	return func(c *config) {
+		c.traceEnabled = traceEnabled
+	}
+}
```
Comment (knusbaum) [inline ddtrace/tracer/option.go:494]:
```suggestion
func WithTraceEnabled(enabled bool) StartOption {
	return func(c *config) {
		c.enabled = enabled
	}
}
```

File: ddtrace/tracer/option_test.go [hunk 3]
```diff
@@ -138,6 +138,22 @@ func TestTracerOptionsDefaults(t *testing.T) {
 		assert.Equal(env, c.env)
 	})
 
+	t.Run("trace_enabled", func(t *testing.T) {
+		t.Run("default", func(t *testing.T) {
+			tracer := newTracer()
+			c := tracer.config
+			assert.True(t, c.traceEnabled)
```
Comment (knusbaum) [inline ddtrace/tracer/option_test.go:145]:
```suggestion
			assert.True(t, c.enabled)
```

File: ddtrace/tracer/option_test.go [hunk 4]
```diff
@@ -138,6 +138,22 @@ func TestTracerOptionsDefaults(t *testing.T) {
...
+		})
+
+		t.Run("override", func(t *testing.T) {
+			os.Setenv("DD_TRACE_ENABLED", "false")
+			defer os.Unsetenv("DD_TRACE_ENABLED")
+			tracer := newTracer()
+			c := tracer.config
+			assert.False(t, c.traceEnabled)
```
Comment (knusbaum) [inline ddtrace/tracer/option_test.go:153]:
```suggestion
			assert.False(t, c.enabled)
```

File: ddtrace/tracer/option_test.go [hunk 5]
```diff
@@ -509,6 +525,30 @@ func TestWithHostname(t *testing.T) {
...
+	})
+
+	t.Run("env-override", func(t *testing.T) {
+		assert := assert.New(t)
+		os.Setenv("DD_TRACE_ENABLED", "false")
+		defer os.Unsetenv("DD_TRACE_ENABLED")
+		c := newConfig(WithTraceEnabled(true))
+		assert.True(c.traceEnabled)
+	})
+}
```
Comment (knusbaum) [inline ddtrace/tracer/option_test.go:550]:
```suggestion
func TestWithTraceEnabled(t *testing.T) {
	t.Run("WithTraceEnabled", func(t *testing.T) {
		assert := assert.New(t)
		c := newConfig(WithTraceEnabled(false))
		assert.False(c.enabled)
	})

	t.Run("env", func(t *testing.T) {
		assert := assert.New(t)
		os.Setenv("DD_TRACE_ENABLED", "false")
		defer os.Unsetenv("DD_TRACE_ENABLED")
		c := newConfig()
		assert.False(c.enabled)
	})

	t.Run("env-override", func(t *testing.T) {
		assert := assert.New(t)
		os.Setenv("DD_TRACE_ENABLED", "false")
		defer os.Unsetenv("DD_TRACE_ENABLED")
		c := newConfig(WithTraceEnabled(true))
		assert.True(c.enabled)
	})
}
```

File: ddtrace/tracer/tracer.go [hunk 6]
```diff
@@ -112,6 +112,9 @@ func Start(opts ...StartOption) {
 		return // mock tracer active
 	}
 	t := newTracer(opts...)
+	if !t.config.traceEnabled {
```
Comment (knusbaum) [inline ddtrace/tracer/tracer.go:115]:
```suggestion
	if !t.config.enabled {
```

**Review by knusbaum — CHANGES_REQUESTED** [review]
Wrong variable name. I'm applying some suggestions.

**Review by knusbaum — APPROVED** [review]

**Pull request merged**